Adds preference for attributes on the same line as what they attribute to the StyleGuide (#63)
wwt/SwiftCurrent: .github/STYLEGUIDE.md

  ```
  </details>

> * **PREFER attributes on the same line as what they are attributing.**

Reviewer: Is this the rule that should also dictate order of attributes?  Is that a rule we need?
```swift
@MainActor @ViewBuilder func createViews() {
```
or
```swift
@ViewBuilder @MainActor func createViews() {
```

Reviewer 2: The order of attributes should be the same kind of rules that we have for other ordering, based off of effect on the overall codebase. So `@MainActor` affects every call site, `@discardableResult` does as well. `@ViewBuilder` only affects that function really, and `@inlineable` is a somewhat transparent compiler optimization.

Reviewer: I think this makes sense.  +1 to Same rules as our file organization

Author: `@MainActor @ViewBuilder func createViews() {` was added as an example. Should the ordering preference be a part of the existing preferences or are we proposing a 3rd addition to call out order of attributes?

Reviewer: I think a 3rd rule is needed for ordering.  It essentially repeats or summarizes the rule for file organization.  The rule holds true despite being in the stacked or inline format.  Items with the most impact come first.

Author: This might be covered by the "Why" of "DO place content in the correct order..." but if we need more clarity I'm on board with either expanding the existing or creating a new "DO" for ordering attributes.

Reviewer: Expanding the existing one would be enough.  Just something to point at that says "Do it this way"

Author: Expanded the examples of the correct ordering guideline.

  <details>

  #### Why?

>     Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning we put behind preferring a return on the same line as the final parameter in a function declaration.

Reviewer 2: "applied to" is fine. I am also not sure I like yours better, fun fact...

>     Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning we put behind preferring a return on the same line as the final parameter in a function declaration.

Author: ```suggestion
    Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning used for preferring a return on the same line as the final parameter in a function declaration.
```

Not sure I like this one any better than "applied to"

>     Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning we put behind preferring a return on the same line as the final parameter in a function declaration.

Reviewer: If we keep this line for splitting into 2 rules, I think this one reads better.

>     Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning we put behind preferring a return on the same line as the final parameter in a function declaration.

Reviewer 3: I like `used for` here

>     NOTE: This does not come at the cost of sanity. If you have an exceptionally large number of attributions then it might indicate that you've gotten overzealous, and it actually obscures your meaning, rather than add to it.

Reviewer 2: Should we call out the `@available` attribute?

Author: Calling out splitting into 2 potential guidelines one for inline when attributes are on functions, properties, or closures and another for attribute on the line above for classes, structs, and actors.

`@available` is still an outlier in that it always goes on its own line, so yes `@available` makes sense to be called out specifically at least in regards to inline preference.

Reviewer: I think 2 rules make sense.  We should also show an example of what they look like with multiple attributes so that I know if I'm inlining on a separate line or stacking them.  And I think `@available` is an exception entirely and is always on its own line by itself.

  ```swift
  // WRONG
  @objc
  func thingExposedToObjectiveC() {
    // ...
  }
  
  // WRONG

>   @ViewBuilder @discardableResult @MainActor @inlinable func betYouDoNotEvenKnowWhatThisFunctionDoesAnymore() {

Author: This one should be clearer that what we mean is "You're doing something wrong if you have this many attributes" rather than "Split them up on separate lines"

>   @ViewBuilder @discardableResult @MainActor @inlinable func betYouDoNotEvenKnowWhatThisFunctionDoesAnymore() {

Reviewer 2: Part of me wants to say that many attributes on a function or property are simply a smell and since this is a "PREFER" I'm tempted to say don't worry about it?

That being said I am not at all a fan of 1 attribute per line, I much prefer the logical breakdown of related attributes, if possible.

>   @ViewBuilder @discardableResult @MainActor @inlinable func betYouDoNotEvenKnowWhatThisFunctionDoesAnymore() {

Reviewer: I think I'm looking for an example of what is something that is readable but still long.  And maybe a why...

```swift
@DependencyInjected @objc @CustomWorkflowAttribute var variableName: String?
```

>   @ViewBuilder @discardableResult @MainActor @inlinable func betYouDoNotEvenKnowWhatThisFunctionDoesAnymore() {

Author: Added an example with 2 attributes on the same line.

  ```
  </details>

* **PREFER attributes on the same line as what they are attributing for properties, functions, and closures.**

  <details>

  #### Why?

    Because we naturally read lines of code as a consistent line, introducing whitespace makes it that much harder to understand the full context of the line you're trying to understand. This is the same reasoning used for preferring a return on the same line as the final parameter in a function declaration. A notable exception to this preference is the `@available` attribute which should always be on its own line.

    NOTE: This does not come at the cost of sanity. If you have an exceptionally large number of attributions then it might indicate that you've gotten overzealous, and it actually obscures your meaning, rather than add to it.

  ```swift
  // WRONG
  @objc
  func thingExposedToObjectiveC() {
    // ...
  }
  
  // WRONG
  @ViewBuilder @discardableResult @MainActor @inlinable func betYouDoNotEvenKnowWhatThisFunctionDoesAnymore() {
    // ...
  }

  // WRONG
  @available(iOS 14.0, *) func doSomething() {
    // ...
  }

  // RIGHT
  @State var stateVar: Bool = true
  @discardableResult func thing() -> Bool { true }

>   @MainActor @ViewBuilder func createViews() {

Reviewer: ```suggestion
  @MainActor @ViewBuilder func createViews() -> some View {
```

    // ...
  }

  // RIGHT
  @available(iOS 14.0, *) 
  func doSomething() {
    // ...
  }
  ```
  </details>

* **PREFER attributes on a separate line from what they are attributing for classes, structs, and actors.**

  <details>

  #### Why?
  This is the convention set by Apple in the [Swift attribute docs](https://docs.swift.org/swift-book/ReferenceManual/Attributes.html). A notable exception to this preference is the `@available` attribute which should always be on its own line.
    

  ```swift
  // WRONG
  @dynamicCallable class MyClass {
    // ...
  }

  // WRONG
  @MainActor struct MyStruct {
    // ...
  }

  // WRONG
  @available(iOS 14.0, *) @MainActor 
  struct MyStruct {
    // ...
  }
  
  // RIGHT

>   @dynamicCallable class MyClass {

Reviewer: ```suggestion
  @dynamicCallable
  class MyClass {
```
